File: javascript/1/2.js
function permutations (str1, str2) {
  const hash = {};

  for (let i = 0; i < str1.length; i++) {
    if (str1[i] in hash) {
      hash[str1[i]] += 1;
    } else {
      hash[str1[i]] = 1;
    }
  }

  for (let j = 0; j < str2.length; j++) {
    if (!(str2[j] in hash)) return false;
    if (hash[str2[j]] === 0) return false; // can't be negative
    hash[str2[j]] -= 1;
  }

  return !Object.values(hash).some(val => val !== 0);
}

File: ruby/1/2.rb
# see if two strings are permutations of each other

# O(n^2) worst case as ruby uses quick sort (n is length of bigger string)
def permutations?(str1, str2)
  return false if str1.size != str2.size
  str1.chars.sort == str2.chars.sort
end

def permutations?(str1, str2)
  return false if str1.size != str2.size
  hash = Hash.new(0);

  str1.each_char do |char|
    hash[char] += 1
  end

  str2.each_char do |char|
    hash[char] -= 1
  end

  !hash.values.any? { |val| val != 0 }
end